### PR TITLE
recreate ES/kibana certs when DNS names change

### PR DESCRIPTION
**Pull request by lmm**
## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/logstorage/logstorage_controller.go [hunk 1]
```diff
@@ -570,29 +573,78 @@ func (r *ReconcileLogStorage) Reconcile(request reconcile.Request) (reconcile.Re
...
-	if err := r.client.Get(ctx, types.NamespacedName{Name: render.TigeraElasticsearchCertSecret, Namespace: render.OperatorNamespace()}, secret); err != nil {
-		if errors.IsNotFound(err) {
-			svcDNSNames := dns.GetServiceDNSNames(render.ElasticsearchServiceName, render.ElasticsearchNamespace, r.clusterDomain)
-			secret, err = render.CreateOperatorTLSSecret(nil,
-				render.TigeraElasticsearchCertSecret, "tls.key", "tls.crt",
-				render.DefaultCertificateDuration, nil, svcDNSNames...,
-			)
-		} else {
+
+	createCert := func() (*corev1.Secret, error) {
```
Comment (Brian-McM) [inline pkg/controller/logstorage/logstorage_controller.go:579]: I don't think we should use an anonymous function here, if it's useful to do this here it'll be useful else where and should be a function that can be accessed by all.

File: pkg/controller/logstorage/logstorage_controller.go [hunk 2]
```diff
@@ -570,29 +573,78 @@ func (r *ReconcileLogStorage) Reconcile(request reconcile.Request) (reconcile.Re
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileLogStorage) elasticsearchSecrets(ctx context.Context) ([]*corev1.Secret, error) {
-	var secrets []*corev1.Secret
+func (r *ReconcileLogStorage) getCertificateKeySecret(ctx context.Context, keySecretName string, keySecretNs string, pubSecretName string, pubSecretNs string, svcDNSNames ...string) (*corev1.Secret, error) {
```
Comment (Brian-McM) [inline pkg/controller/logstorage/logstorage_controller.go:576]: I don't think this function belongs to the logstorage reconciler, could you please put this in a suitable common location.

**Review by Brian-McM — CHANGES_REQUESTED** [review]

File: pkg/controller/utils/elasticsearch.go [hunk 3]
```diff
@@ -20,8 +20,10 @@ import (
 	"crypto/tls"
```
Comment (Brian-McM) [inline pkg/controller/utils/elasticsearch.go:20]: I guess this is a bit nitpicky, but I don't think this has anything to do with elasticsearch so I don't think it should be in this file.

File: pkg/controller/utils/elasticsearch.go [hunk 4]
```diff
@@ -373,3 +376,75 @@ func getTotalEsDisk(ls *operatorv1.LogStorage) int64 {
...
+	pemBlock, _ := pem.Decode(certBytes)
+	cert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return false, err
+	}
+
+	return reflect.DeepEqual(cert.DNSNames, expectedDNSNames), nil
+}
+
+func GetCertificateSecret(ctx context.Context, client client.Client, name string, ns string) (*corev1.Secret, error) {
```
Comment (Brian-McM) [inline pkg/controller/utils/elasticsearch.go:441]: This doesn't really have anything to do with specifically getting a secret that contains a certificate, it's just shortcutting the code to get a secret in general.

Comment (lmm) [reply pkg/controller/utils/elasticsearch.go]: Moved to its own file

File: pkg/controller/utils/elasticsearch.go (hunk 3 above)
Comment (lmm) [reply pkg/controller/utils/elasticsearch.go]: Makes sense. Moved to a new file

**Review by doublek — APPROVED** [review]
LGTM

**Comment by lmm** [comment]
After discussing with @doublek , I'll add a follow-up PR tomorrow

**Pull request merged**